packages/transaction-pay-controller/src/utils/required-tokens.ts

```diff
@@ -0,0 +1,280 @@
+
+  const hasBalance = new BigNumber(token.balanceRaw).isGreaterThanOrEqualTo(
+    token.amountRaw,
+  );
+
+  if (hasBalance || amountUsdValue.isGreaterThanOrEqualTo(1)) {
+    return {
+      ...token,
+      allowUnderMinimum: true,
+      skipIfBalance: true,
+    };
+  }
+
+  const fiatRates = getTokenFiatRate(
+    messenger,
+    nativeTokenAddress,
+    chainId,
+  ) as FiatRates;
+
+  const oneDollarRawHex = add0x(
+    new BigNumber(1).dividedBy(fiatRates.usdRate).shiftedBy(18).toString(16),
+  );
+
+  const oneDollarToken = buildRequiredToken(
+    transaction,
+    nativeTokenAddress,
+    oneDollarRawHex,
+    messenger,
+  );
+
+  /* istanbul ignore next */
+  if (!oneDollarToken) {
+    return undefined;
+  }
+
+  return {
+    ...oneDollarToken,
+    allowUnderMinimum: true,
+    skipIfBalance: true,
+  };
+}
+
+/**
+ * Get the full token properties for a specific token and amount.
+ *
+ * @param transaction - Transaction metadata.
+ * @param tokenAddress - Token address.
+ * @param amountRawHex - Raw token amount in hexadecimal format.
+ * @param messenger - Controller messenger.
+ * @returns The full token properties or undefined if the token data could not be retrieved.
+ */
+function buildRequiredToken(
+  transaction: TransactionMeta,
+  tokenAddress: Hex,
+  amountRawHex: Hex,
+  messenger: TransactionPayControllerMessenger,
+): TransactionPayRequiredToken | undefined {
+  const { chainId, txParams } = transaction;
+  const from = txParams.from as Hex;
+
+  const { decimals: tokenDecimals, symbol } =
+    getTokenInfo(messenger, tokenAddress, chainId) ?? {};
+
+  const fiatRates = getTokenFiatRate(messenger, tokenAddress, chainId);
+  const tokenBalance = getTokenBalance(messenger, from, chainId, tokenAddress);
+
+  if (tokenDecimals === undefined || !symbol || fiatRates === undefined) {
+    return undefined;
+  }
+
+  const {
+    amountHuman: balanceHuman,
+    amountRaw: balanceRaw,
+    amountFiat: balanceFiat,
+    amountUsd: balanceUsd,
+  } = calculateAmounts(tokenBalance, tokenDecimals, fiatRates);
+
+  const { amountHuman, amountRaw, amountFiat, amountUsd } = calculateAmounts(
+    amountRawHex,
+    tokenDecimals,
+    fiatRates,
+  );
+
+  return {
+    address: tokenAddress,
+    allowUnderMinimum: false,
+    amountFiat,
+    amountHuman,
+    amountRaw,
+    amountUsd,
+    balanceFiat,
+    balanceHuman,
+    balanceRaw,
+    balanceUsd,
+    chainId,
+    decimals: tokenDecimals,
+    skipIfBalance: false,
+    symbol,
+  };
+}
+
+/**
+ * Calculates the various amount representations for a token value.
+ *
+ * @param amountRawInput - Raw amount.
+ * @param decimals - Number of decimals for the token.
+ * @param fiatRates - Fiat rates for the token.
+ * @returns Object containing amount in fiat, human-readable, raw, and USD formats.
+ */
+function calculateAmounts(
+  amountRawInput: BigNumber.Value,
+  decimals: number,
+  fiatRates: FiatRates,
+) {
+  const amountRawValue = new BigNumber(amountRawInput);
```

Comment:
do we need to pass 16 here to denote a hexadecimal value?

Reply:
If the value starts with `0x`, it should be automatic.